### PR TITLE
Distinguish never-spawned from never-submitted.

### LINKED ISSUE #6066: suicide triggers broken
This example works (tasks a and c automatically removed, workflow shuts down) in Cylc 7 but stalls in Cylc 8:

```ini
# suite.rc
[scheduling]
    cycling mode = integer
    initial cycle point = 1
    [[dependencies]]
        [[[R1]]]
            graph = """
                a & b => c
                a:fail => !a & !c
            """

[runtime]
    [[a]]
        script = false
    [[b,c]]
```

This example also stalls in Cylc 8:

```ini
# flow.cylc
[scheduling]
    cycling mode = integer
    initial cycle point = 1
    [[dependencies]]
        [[[R1]]]
            graph = """
                a? & b => c
                a:fail? => !a & !c
            """

[runtime]
    [[a]]
        script = false
    [[b,c]]
```

Slightly related to https://github.com/cylc/cylc-flow/issues/5729 as this covers some or all of the remaining use cases for suicide triggers.

### PR DESCRIPTION
**Pull request by hjoliver**
Close #6066

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by hjoliver** [comment]
Fixing [bullet point 2](https://github.com/cylc/cylc-flow/issues/6066#issuecomment-2054228526) has generated a DB issue that I don't understand yet.

File: cylc/flow/task_pool.py
Comment (oliver-sanders) [inline cylc/flow/task_pool.py:1653]: "Never spawned" doesn't necessarily mean "suicided", I think this PR has implications on `cylc remove` too. Presumably, this will prevent tasks being reincarnated after being manually removed (as well as after being automatically suicided)?

This might be a positive behaviour change for `cylc remove`, but worth noting that is may clash with #5643.

Comment (hjoliver) [reply cylc/flow/task_pool.py]: (Follow-up on main thread)

**Comment by hjoliver** [comment]
### suicide trigger

```ini
[scheduling]
    [[graph]]
        R1 = """
            a? & b => c
            a:failed? => !a & !c
        """
[runtime]
    [[a]]
        script = "sleep 3; false"
    [[b,c]]
```
now works as expected:
```console
...
INFO - [1/c:waiting(runahead)] => waiting
INFO - [1/a/01:failed] removed from active task pool: suicide trigger
INFO - [1/c:waiting] removed from active task pool: suicide trigger
INFO - [1/b/01:running] => succeeded
INFO - Not spawning 1/c: already used in this flow
INFO - [1/b/01:succeeded] removed from active task pool: completed
INFO - Workflow shutting down - AUTOMATIC
INFO - DONE
```

--------
## cylc remove

Since #5643 is not done yet, we can't remove future tasks, so need an example that spawns a waiting task to be removed:

```ini
[scheduling]
    [[graph]]
        # b spawns c before a removes c
        R1 = "a & b => c"
[runtime]
    [[a]]
        script = "sleep 5; cylc remove $CYLC_WORKFLOW_ID//1/c"
    [[b,c]]
```

Result: `1/c` does not run (already spawned in this flow; and prior removal was logged).

```console
...
INFO - [1/c:waiting(runahead)] => waiting
INFO - [1/b/01:succeeded] removed from active task pool: completed
INFO - Command "remove_tasks" received. ID=f2834293-11dc-49cd-ae5c-e1d9f5cbdb42
    remove_tasks(tasks=['1/c'])
INFO - [1/c:waiting] removed from active task pool: request
INFO - Command "remove_tasks" actioned. ID=f2834293-11dc-49cd-ae5c-e1d9f5cbdb42
INFO - [1/a/01:running] => succeeded
INFO - Not spawning 1/c: already used in this flow
INFO - [1/a/01:succeeded] removed from active task pool: completed
INFO - Workflow shutting down - AUTOMATIC
INFO - DONE
```

I think this is what we want, subject to possible revision when #5643 is implemented.

**Comment by hjoliver** [comment]
Can rebase this onto 8.2.x if necessary (it seems we may get another 8.2 release out before 8.3.0?)

**Comment by oliver-sanders** [comment]
Yes, we've got a cylc-rose issue that we would like to fix in the 8.2 series so are planning another release.

**Comment by hjoliver** [comment]
This actually conflicts quite badly with 8.2.x. Mind if we leave it on master 8.3.0?

**Comment by oliver-sanders** [comment]
Shockingly, we've only had one report of this so far (failure cases are presumably less well tested in workflows?) so we can wait a little.

**Comment by oliver-sanders** [comment]
Looks good to me, I wrote a couple of extra tests whilst reviewing, can confirm that this fixes the bug where removed tasks could be automatically respawned at a later time by task pool logic. There's no associated issue for this, but might be worth a mention in the changelog entry.

https://github.com/hjoliver/cylc-flow/pull/49

**Comment by hjoliver** [comment]
> I wrote a couple of extra tests whilst reviewing - https://github.com/hjoliver/cylc-flow/pull/49

Thanks, I'll take a look tomorrow. I presume you saw my new integration tests on this branch (which seem to achieve 100% patch coverage) - do yours go beyond mine?

**Comment by oliver-sanders** [comment]
Yes, I extended your test to ensure that it's possible to get the task back by triggering and added a new one to ensure that removed tasks remain removed.

**Review by oliver-sanders — APPROVED** [review]

File: cylc/flow/task_pool.py
Comment (hjoliver) [reply cylc/flow/task_pool.py:1657]: Indeed, but how to describe what's happening concisely? 

Basically, the flow (on this branch of the graph) came to a halt at this point because the task was removed earlier by suicide trigger or `cylc remove`. That definitely warrants some kind of log message.

I'll change it to this (subject to further litigation, of course):
```suggestion
            LOG.info(f"Flow stopping at {point}/{name} - task previously removed")
```

File: changes.d/6067.fix.md
Comment (hjoliver) [reply changes.d/6067.fix.md:1]: Another tricky terminology issue to sort out 😬  We don't typically (well, never) use "the workflow" to mean just the scheduler task pool, so what will "added back into the workflow" mean to users?

I'll change it to this for now, to avoid saying exactly what the task is "added back" to: 
```suggestion
Fixed a bug that sometimes allowed suicide-triggered or manually removed tasks to be added back later.
```

File: cylc/flow/task_pool.py
Comment (hjoliver) [reply cylc/flow/task_pool.py:842]: This was to go with "added to active task pool", which Ronnie has also commented on below.

I'll demote both to DEBUG for now, but we need to discuss it a bit in the chat rooms...

**Comment by hjoliver** [comment]
All comments addressed. Logging is nice and succinct now:
```
$ cylc play --no-detach --no-timestamp bug/run17

 ▪ ■  Cylc Workflow Engine 8.3.0.dev
 ██   Copyright (C) 2008-2024 NIWA
▝▘    & British Crown (Met Office) & Contributors

INFO - Extracting job.sh to /home/oliverh/cylc-run/bug/run17/.service/etc/job.sh
INFO - Workflow: bug/run17
INFO - Scheduler: url=tcp://NIWA-1022450.niwa.local:43023 pid=18085
INFO - Workflow publisher: url=tcp://NIWA-1022450.niwa.local:43032
INFO - Run: (re)start number=1, log rollover=1
INFO - Cylc version: 8.3.0.dev
INFO - Run mode: live
INFO - Initial point: 1
INFO - Final point: 1
INFO - Cold start from 1
INFO - New flow: 1 (original flow from 1) 2024-04-23T13:40:13
INFO - [1/a:waiting(runahead)] => waiting
INFO - [1/a:waiting] => waiting(queued)
INFO - [1/a:waiting(queued)] => waiting
INFO - [1/a:waiting] => preparing
INFO - [1/a/01:preparing] submitted to localhost:background[18110]
INFO - [1/a/01:preparing] => submitted
INFO - [1/a/01:submitted] => running
INFO - [1/a/01:running] => succeeded
INFO - Workflow shutting down - AUTOMATIC
INFO - DONE
```

File: cylc/flow/task_pool.py
Comment (MetRonnie) [reply cylc/flow/task_pool.py]: I don't like "flow blocked" I'm afraid! How about:
```
Skipping {point}/{name} {stringify_flow_nums(flow_nums)} - task previously removed
```

Comment (hjoliver) [reply cylc/flow/task_pool.py]: I think "flow blocked" is better than your two suggestions, because it correctly indicates that the flow will not continue here because the task was removed.

"Skipping" suggests the flow will skip over the remove task, which is not the case.

"Spawning" (as you'll have seen in the chat) is I think being expunged from user terminology.

Comment (oliver-sanders) [reply cylc/flow/task_pool.py]: Anther terminology debate :facepalm:. FYI, I don't think that any of these suggestions will make a lick of sense to users.

"Flow Blocked" is incorrect, the flow isn't "blocked", it's most likely continuing elsewhere in the graph, it just isn't proceeding through this particular task by design.

"Skipping" is incorrect as the flow won't "skip" over the task.

"Not Spawning" is correct (or not *re*-spawning if we're being picky). Note, the term "spawn" is not being "expunged", it was never a user facing term in the first place.

This log message isn't really relevant to users, the removal / suicide was already logged in a way that they can understand, this is an engineers message for debugging the downstream consequences of removal, it is only relevant if the removal mechanism doesn't work for some reason. So we can call it whatever we like and stick it at debug level.

Comment (hjoliver) [reply cylc/flow/task_pool.py]: >  FYI, I don't think that any of these suggestions will make a lick of sense to users.

I agree, mostly, but unfortunately that doesn't absolve us from coming up with consistent terminology for these behaviours, and documenting it so that users *can* understand it if they want to or need to.

> "Flow Blocked" is incorrect, the flow isn't "blocked", it's most likely continuing elsewhere in the graph, it just isn't proceeding through this particular task by design.

It is correct in the sense that I meant it: the flow is literally blocked *on that branch of the graph*, it doesn't mean the entire flow is blocked or stopped.

> "Skipping" is incorrect as the flow won't "skip" over the task.

Agreed.

> Not Spawning" is correct (or not re-spawning if we're being picky).

Agreed, but  you've been arguing against using "spawn" as a user-facing term anymore on Element chat, and the scheduler log IS user-facing. Users often need to look at the look to figure out what their workflow did, or how it ended up in a particular state.

>  Note, the term "spawn" is not being "expunged", it was never a user facing term in the first place.

Ah, it was, and still is. e.g.:
 - the long-standing pre-Cylc 8 config, `spawn to max active cycle points`
 - in the (user-facing!) scheduler log
 - and in the user guide, where it describes how the workflow advances
 
The word "spawn" is used a bunch of times in the user guide "concurrent flows" section for instance.

In fact, reading that again made me realize that "spawn" is that right user-facing term for that concept. E.g (User Guide) *This triggers a task with no flow numbers. It will not spawn children, and other flows that come by will re-run it.*

> This log message isn't really relevant to users, the removal / suicide was already logged in a way that they can understand, this is an engineers message for debugging the downstream consequences of removal, it is only relevant if the removal mechanism doesn't work for some reason. So we can call it whatever we like and stick it at debug level.
 
 I disagree with this. The removal could have been done, and logged, quite some time ago, and it could have been done automatically by suicide trigger or by scripted use of `cylc remove`. We need to log again, at INFO level, why the flow (on this branch of the graph) is stopping at this point in the graph.

Comment (hjoliver) [reply cylc/flow/task_pool.py]: OK, I think options are as follows (or variants thereof):
- *flow blocked here, the task was removed earlier*  (note addition of "here" to avoid implicating the entire flow)
- *not spawning task, it was removed earlier*

Given that @oliver-sanders agrees "spawning" is correct, and that (as I've just shown) spawning probably should *remain* a user-facing term, perhaps we should stick with that.

However, it needs to be logged at INFO level to explain why the flow suddenly halts there (on this branch of the graph).

Comment (oliver-sanders) [reply cylc/flow/task_pool.py]: > Note, the term "spawn" ... was never a user facing term in the first place.
>
>> Ah, really?!

That's Cylc 7!

I'm very confused by these comments. It was your own argument that users should not need to understand the management (insertion or removal) of task proxies or task pool logic in general which I got on board with. It was a primary goal of SoD to remove the need for us to communicate this internal implementation detail, and a necessary side effect as SoD would make it harder for users to monitor the task pool itself. That's why spawning was retired right, `s/spawn/set-outputs`, users shouldn't need to understand spawning, because SoD will ensure the task proxy is always there when needed right? E.G. https://github.com/cylc/cylc-flow/issues/2143. I agreed with you on this at the time, it was a good idea!

I appreciate this is going back a few years now, maybe reality crept in and the thinking has changed. Perhaps it was a bit ambitions to presume we could hide the internal logic. Maybe we should write up the SoD model and communicate it to users as essential reading?


> It needs to be logged at INFO level though, to explain why the flow does not continue here, at this time.

I'm not convinced it does. We are trying to communicate to users the graph model, not the internal SoD implementation (or at least I thought we were).

Take this graph for example:

```
a => b => c => d
a => x
b => x
c => x
d => x
```

If the user runs `cylc remove //x`, they will get a message along the lines of "x removed".

They don't need to see four "not re-spawning x" messages in a row as the graph proceeds, there is no new information in these messages. Moreover, it isn't communicating anything to them about the graph. Of course x isn't be re-spawned, I removed it! On purpose!

Comment (oliver-sanders) [reply cylc/flow/task_pool.py]: My approval still stands of course, I'll leave you two to find something you can agree on.

I was trying to overt a terminology argument not to stoke one, I was ok with the original suggestion.

Comment (MetRonnie) [reply cylc/flow/task_pool.py]: I just think "flow blocked" sounds too negative and some users will see it and think something has gone wrong

**Comment by hjoliver** [comment]
> To quote [your own proposal](https://github.com/cylc/cylc-admin/blob/master/docs/proposal-spawn-on-d.md#advantages):
> 
>>    Users should not have to know about the "task pool" and the concepts of task "insert" and "remove" anymore.

Ah, OK @oliver-sanders, sorry, I do understand your point now.  

That statement was somewhat idealistic in retrospect. And maybe that's my fault. But it is kinda true in principle for Cylc 8 now, in the following sense:
- the content of the "task pool", unlike in Cylc 7, can now be easily explained in conceptual terms (try doing that for Cylc 7)
- users never need to manually "insert" task proxies into the pool, that now happens automatically as a result of conceptual operations like "trigger"

HOWEVER, despite being much more sensible and explainable now, the content of the task pool is still non-trivial (it doesn't just contain active tasks), and it the n=0 basis of the GUI window, so we have to call it something, document what's in it, and how tasks enter and exit the pool.  After going through all the options that we've thought of (to my knowledge) it still seems to me that "task pool" and "completed outputs spawn downstream tasks" is the best way to describe that.

**Comment by hjoliver** [comment]
> If the user runs cylc remove //x, they will get a message along the lines of "x removed".
>
> They don't need to see four "not re-spawning x" messages in a row as the graph proceeds, there is no new information in these messages. 

Yes, but you've given a simple example where the removal and its effects are contiguous and obvious. In more complex workflows, and especially once we can remove future (non-pool) tasks as planned, that may not be the case. 

This is not SoD implementation, it just explains why the flow does not continue at this point, in this branch of the graph: because at some point in the past someone or something "removed" a task in the path of the flow.

File: cylc/flow/task_pool.py
Comment (hjoliver) [reply cylc/flow/task_pool.py]: OK @MetRonnie, let's try to agree. I don't see "flow blocked" as negative, but perhaps we could add more information.

We've established that "skipping" isn't right. Are you happy with either of these, or do you have another suggestion?:

- "flow blocked by previous removal of <task>"
- "not spawning <task> (previously removed)"

I like "flow blocked" because it automatically conveys the fact that the flow will not continue from this point.

**Comment by hjoliver** [comment]
Just noticed you approved this @MetRonnie - I could have saved myself some writing above 🤦‍♂️ 

OK, I'll merge this after tweaking the log message to give a bit more information as described above, so we can push on to 8.3.0. 

@oliver-sanders I'll leave it at INFO because:
- task removal is pretty niche, it won't clutter the log
- sometimes the flow hits a removed task long after the removal

We can push it back to DEBUG later if you convince me that the second point isn't relevant.

**Pull request merged**
